bundles/org.eclipse.equinox.p2.core/src/org/eclipse/equinox/internal/p2/core/ProvisioningAgent.java

```diff
@@ -95,10 +95,16 @@ private void checkRunning() {
 	@Override
 	public void registerService(String serviceName, Object service) {
 		checkRunning();
-		agentServices.put(serviceName, service);
 		if (service instanceof IAgentService) {
 			((IAgentService) service).start();
 		}
+		if (agentServices.put(serviceName, service) instanceof IAgentService prevService) {
+			prevService.stop();
+			if (prevService == service) {
+				agentServices.remove(serviceName, prevService);
+				throw new IllegalArgumentException("Double registration"); //$NON-NLS-1$
```

Comment:
Is it anywhere mentioned that this is not allowed? Also it would be more an illegalstate than an illegal argument and the message should be more descriptive.

Reply:
> If double registration was allowed, IAgentService.start() would be invoked twice for the same object.

This is an implementation detail and can of course be avoided, so the question is if anywhere is documented that registering the same service twice is forbidden (at least it was not before)

Reply:
I've extended the Javadoc.